zcash_client_backend/src/fees/common.rs

```diff
@@ -191,58 +246,90 @@ where
-                        available: total_in,
-                        required: (total_in + shortfall).ok_or_else(overflow)?,
-                    })
-                }
-                DustAction::AllowDustChange => TransactionBalance::new(
+    let transparent = net_flows.is_transparent();
+    let total_out_plus_fee_without_change =
+        (total_out + fee_without_change).ok_or_else(overflow)?;
+    let total_out_plus_fee_with_change = (total_out + fee_with_change).ok_or_else(overflow)?;
+
+    let (change, fee) = {
+        if transparent && total_in < total_out_plus_fee_without_change {
+            // Case 1 for a tx with all transparent flows.
+            return Err(ChangeError::InsufficientFunds {
+                available: total_in,
+                required: total_out_plus_fee_without_change,
+            });
+        } else if transparent && total_in == total_out_plus_fee_without_change {
+            // Case 2 for a tx with all transparent flows.
+            (vec![], fee_without_change)
+        } else if total_in < total_out_plus_fee_with_change {
+            // Case 3a, or case 1 or 2 with non-transparent flows.
+            return Err(ChangeError::InsufficientFunds {
+                available: total_in,
+                required: total_out_plus_fee_with_change,
+            });
+        } else {
+            // Case 3b or 3c.
+            let proposed_change =
+                (total_in - total_out_plus_fee_with_change).expect("checked above");
+            let simple_case = || {
+                (
                     vec![ChangeValue::shielded(
                         change_pool,
                         proposed_change,
                         change_memo,
                     )],
-                    fee_amount,
-                )
-                .map_err(|_| overflow()),
-                DustAction::AddDustToFee => TransactionBalance::new(
-                    vec![],
-                    (fee_amount + proposed_change).ok_or_else(overflow)?,
+                    fee_with_change,
                 )
-                .map_err(|_| overflow()),
+            };
+
+            let dust_threshold = dust_output_policy
+                .dust_threshold()
+                .unwrap_or(default_dust_threshold);
+
+            if proposed_change < dust_threshold {
+                match dust_output_policy.action() {
+                    DustAction::Reject => {
+                        let shortfall = (dust_threshold - proposed_change).ok_or_else(underflow)?;
+
+                        return Err(ChangeError::InsufficientFunds {
+                            available: total_in,
+                            required: (total_in + shortfall).ok_or_else(overflow)?,
+                        });
+                    }
+                    DustAction::AllowDustChange => simple_case(),
+                    DustAction::AddDustToFee => {
+                        let fee_with_dust = (total_in - total_out).expect("can't happen");
```

Comment:
```suggestion
                       let fee_with_dust = (total_in - total_out).unwrap();
```
or else provide a better expectation message than "can't happen"

Reply:
Please move this comment to #1430.